tools/dds/dds-sniffer/rs-dds-sniffer.cpp

```diff
@@ -181,88 +208,175 @@ void dds_sniffer::on_writer_removed( librealsense::dds::dds_guid guid, const cha
     }
 
-    std::lock_guard<std::mutex> lock( _dds_entities_lock );
+    std::lock_guard< std::mutex > lock( _dds_entities_lock );
 
     _discovered_participants.erase( guid );
 }
 
-void dds_sniffer::save_topic_writer( librealsense::dds::dds_guid guid, const char* topic_name )
+void dds_sniffer::on_type_discovery( char const * topic_name, DynamicType_ptr dyn_type )
+{
+    // Register type with participant
+    TypeSupport type_support( new DynamicPubSubType( dyn_type ) );
+    type_support.register_type( _participant.get() );
+    std::cout << "Discovered topic " << topic_name << " of type: " << type_support->getName() << std::endl;
+
+    if (_print_topic_samples)
+    {
+        // Create subscriber, topic and reader to receive instances of this topic
+        if (_discovered_types_subscriber == nullptr)
+        {
+            _discovered_types_subscriber = _participant.get()->create_subscriber( SUBSCRIBER_QOS_DEFAULT, nullptr );
+            if (_discovered_types_subscriber == nullptr)
+            {
+                std::cout << "Cannot create subscriber for discovered type '" << topic_name << std::endl;
+                return;
+            }
+        }
+
+        Topic* topic = _participant.get()->create_topic( topic_name, type_support->getName(), TOPIC_QOS_DEFAULT );
+        if (topic == nullptr)
+        {
+            std::cout << "Cannot create topic for discovered type '" << topic_name << std::endl;
+            return;
+        }
+
+        StatusMask sub_mask = StatusMask::subscription_matched() << StatusMask::data_available();
+        DataReader* reader = _discovered_types_subscriber->create_datareader( topic, DATAREADER_QOS_DEFAULT,
+                                                                              &_reader_listener, sub_mask );
+        if (reader == nullptr)
+        {
+            std::cout << "Cannot create reader for discovered type '" << topic_name << std::endl;
+            _participant.get()->delete_topic( topic );
+            return;
+        }
+        _discovered_types_readers[reader] = topic;
+
+        DynamicData_ptr data( DynamicDataFactory::get_instance()->create_data( dyn_type ) );
+        _discovered_types_datas[reader] = data;
+    }
+}
+
+dds_sniffer::dds_reader_listener::dds_reader_listener( std::map< DataReader *, DynamicData_ptr > & datas )
+    : _datas( datas )
+{
+}
+
+void dds_sniffer::dds_reader_listener::on_data_available( DataReader * reader )
+{
+    const TopicDescription * topic_desc = reader->get_topicdescription();
+    std::cout << "Received topic " << topic_desc->get_name() << " of type " << topic_desc->get_type_name() << std::endl;
+
+    auto dit = _datas.find( reader );
+
+    if( dit != _datas.end() )
+    {
+        DynamicData_ptr data = dit->second;
+        SampleInfo info;
+        if( reader->take_next_sample( data.get(), &info ) == ReturnCode_t::RETCODE_OK )
+        {
+            if( info.instance_state == ALIVE_INSTANCE_STATE )
```

Comment:
No need to check `if( info.valid_data )` ?
As I understood this needs to be checked when reading messages.
Sometime they check it and sometimes they check `if( info.instance_state == ALIVE_INSTANCE_STATE )` 
on their examples, do you know when to check what?

Reply:
Good point...

`info.instance_state == ALIVE_INSTANCE_STATE` means that the writer is still alive and did not dispose of this sample
`info.valid_data == true` means that the sample contains a change in the value. It will be false for changes in the status only.

I think that using `info.valid_data` is better in the sniffer application, but since I `take` the sample not just `read` it, it probably does not matter, the sample and it's state won't be updated.

Reply:
I understand your point.
I think it's better that we align on `is_valid` usage, it's more strait forward and also at this stage we don't care about the writer, if he was alive when he sent the message we want to display it right?

Reply:
Done